### ISSUE TITLE
[meta.logical] unclear wording

### ISSUE THREAD

**Issue by FrankHB** [cplusplus/draft#2548]
In [meta.logical]/2
> then instantiating `conjunction<B1, ..., BN>::value` does not require the instantiation of `Bj::value` for `j > i`

What is "the instantiation of `Bj::value`"?

(The font of `j` should also be fixed.)

The change in #1193 (LWG 2567) also introduce another (possibly non-editorial) problem: does the explicit cast to `bool` of `value` mandated? For example:

```
#include <type_traits>

struct B
{
  constexpr explicit operator bool() const noexcept {return true;}
};

struct C
{
  static constexpr const B value = B();
};

int main()
{
  static_assert(bool(std::conjunction<C, std::true_type>::value));
}

```
Is the above program well-formed?

It seems intentionally ill-formed as per "shall have a member value which is convertible to `bool`" wording in [meta.logical]/4. (Is this presumely "implicit", or like `std::is_convertible`?) Then `bool(Bj::value)` is not appropriate as it suggests the cast shall be supported, as [implemented in cppreference](https://en.cppreference.com/w/cpp/types/conjunction).

Otherwise, libstdc++, libc++ and MSVC currently all has the bug not supporting resoultion of LWG 2567/2568, and [meta.logical]/4 should be reworded to clarify the intention. I guess this is not intentional because a) it is not proposed at first; b) the context here is more like operands of `&&` but not `if`, which needs no contextually conversion, so omission of explicit cast should not be encouraged.

**Comment by jwakely** [comment]
> What is "the instantiation of Bj::value"?

It seems clear to me.

Maybe it would be more precise to say "does not require the definition of the static data member `Bj::value` to exist" with a cross-reference to [temp.inst].

**Comment by jwakely** [comment]
> (The font of j should also be fixed.)

I'm not sure what you mean by "fixed" but I've submitted #2549 to use math mode for the subscripts _i_, _j_, _1_ and _N_.

> The change in #1133 (LWG 2567) also introduces another (possibly non-editorial) problem: does the explicit cast to bool of value mandated?

This seems non-editorial to me.

> Is the program above well-formed?

It's undefined, because it fails to meet the _Requires:_ precondition.

> Then bool(Bj::value) is not appropriate as it suggests the cast shall be supported, as implemented in cppreference.

The cast is supported, but it's not required. If an implementation wants to rely on an implicit conversion instead that's fine (as long as it does so without a narrowing conversion, as per lwg 2587).

> Otherwise, libstdc++, libc++ and MSVC currently all has the bug concerned with the resolution of LWG 2567

That's not an editorial issue with the standard.

**Comment by jwakely** [comment]
> I don't see any precondition prefixed Requires: in this subclause.

Right. It's still a requirement though, and your program doesn't meet it. Please take non-editorial questions to the right place.

**Comment by FrankHB** [comment]
OK. Enlightened by the clauses about the editorial style, I think now I'm clear what happens.

As per [structure.specifications], _Requires:_ is for "descriptions of function semantics" so I guess it is not fit here. Instead [structure.requirements] is used. As a result, [res.on.requirements] takes effect instead of [res.on.required]. Thus, the program is actually ill-formed without diagnostics required, but not undefined.

**Comment by timsong-cpp** [comment]
> The cast is supported, but it's not required. If an implementation wants to rely on an implicit conversion instead that's fine (as long as it does so without a narrowing conversion, as per lwg 2587).

My understanding is that you have to cast: implicit and explicit conversion could yield different results, and the cast was intended to pin down the behavior.

I would not be opposed to saying that "the type of `Bi::value` shall model `ConvertibleTo<bool>`" though (which imposes a requirement that the two conversions do the same thing). But this is not the right place for that discussion.
